### PR TITLE
Exclude default config value from the pip-compile header

### PR DESCRIPTION
**Pull request by atugushev**
Resolves https://github.com/jazzband/pip-tools/pull/1863#issuecomment-1611728125

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Assure PR title is short, clear, and good to be included in the user-oriented changelog

##### Maintainer checklist

- [x] Assure one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

File: piptools/utils.py
Comment (webknjaz) [reply piptools/utils.py:631]: If this is for testing, the resulting callable contains a `__wrapped__` attribute and you could make use of it with `monkeypatch.setattr` in tests instead of dropping the cache in normal runtime.

File: tests/test_utils.py
Comment (webknjaz) [inline tests/test_utils.py:393]: Does pip-sync produce these headers too? I'd add it to the tests, then.

Comment (atugushev) [reply tests/test_utils.py:393]: Nope, pip-sync does not use `get_compile_command`.

File: piptools/utils.py
Comment (atugushev) [reply piptools/utils.py:631]: The cache causes tests to become flaky and fail, depending on their ordering. I wonder why this is cached in the first place, especially as this function is only called once.

**Comment by chrysle** [comment]
Looks like something is wrong with QA & CI.. Let's restart the jobs tomorrow.  And maybe wait until #1894 is merged for the failing PyPy tests.

File: piptools/utils.py
Comment (webknjaz) [reply piptools/utils.py:631]: Okay, fair.

Comment (webknjaz) [reply piptools/utils.py:631]: Marking this as resolved.

File: tests/test_utils.py
Comment (webknjaz) [reply tests/test_utils.py:393]: Marking as resolved.

**Review by webknjaz — APPROVED** [review]

**Comment by webknjaz** [comment]
This will need rebasing on top of https://github.com/jazzband/pip-tools/pull/1895.

**Comment by q0w** [comment]
@atugushev There is a bug with pip 6.14.0.
I just use `pip-compile --allow-unsafe --generate-hashes --output-file=test.txt --resolver=backtracking --strip-extras test.in`
but now it adds `--config=pyproject.toml` even there is no configuration in pyproject.toml
When running pip-compile click context has no src_files, so select_config_file returns pyproject.toml, but after compiling click context has src_files as test.in, so select_config_file returns None and so pyproject.toml != None

Making `--config` is_eager=False works

**Comment by atugushev** [comment]
@q0w thanks for heads up. Referenced this in https://github.com/jazzband/pip-tools/issues/1903

**Pull request merged**
